File: java/common/foreign-bridge/src/main/java/sleeper/foreign/datafusion/FFISleeperOptions.java
/*
 * Copyright 2022-2026 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.foreign.datafusion;

import jnr.ffi.Struct;

/**
 * Contains all the Sleeper options for DataFusion operation. These may not be needed for every DataFusion usage,
 * so come with reasonable defaults.
 *
 * <strong>THIS IS A C COMPATIBLE FFI STRUCT!</strong> If you updated this struct (field ordering, types, etc.),
 * you MUST update the corresponding Rust definition in rust/sleeper_df/src/objects/ffi_sleeper_options.rs. The
 * order and types of the fields must match exactly.
 */
@SuppressWarnings("checkstyle:membername")
public class FFISleeperOptions extends Struct {
    /** Whether we should use readahead when reading from S3. */
    public final Struct.Boolean use_readahead_store = new Struct.Boolean();
    /** Whether Parquet page indexes should be read. */
    public final Struct.Boolean read_page_indexes = new Struct.Boolean();
    /** Maximum size of output Parquet row group in rows. */
    public final Struct.size_t max_row_group_size = new Struct.size_t();
    /** Maximum size of output Parquet page size in bytes. */
    public final Struct.size_t max_page_size = new Struct.size_t();
    /** Output Parquet compression codec. */
    public final Struct.UTF8StringRef compression = new Struct.UTF8StringRef();
    /** Output Parquet writer version. Must be v1 or v2. */
    public final Struct.UTF8StringRef writer_version = new Struct.UTF8StringRef();
    /** Column min/max values truncation length in output Parquet. */
    public final Struct.size_t column_truncate_length = new Struct.size_t();
    /** Max sizeof statistics block in output Parquet. */
    public final Struct.size_t stats_truncate_length = new Struct.size_t();
    /** Should row key fields use dictionary encoding in output Parquet. */
    public final Struct.Boolean dict_enc_row_keys = new Struct.Boolean();
    /** Should sort key fields use dictionary encoding in output Parquet. */
    public final Struct.Boolean dict_enc_sort_keys = new Struct.Boolean();
    /** Should value fields use dictionary encoding in output Parquet. */
    public final Struct.Boolean dict_enc_values = new Struct.Boolean();

    public FFISleeperOptions(jnr.ffi.Runtime runtime) {
        super(runtime);
    }

    /**
     * Validates the state of this struct.
     *
     * @throws IllegalStateException when a invariant fails
     */
    public void validate() {
        if (max_row_group_size.get() < 1) {
            throw new IllegalStateException("max row group size < 1");
        }
        if (max_page_size.get() < 1) {
            throw new IllegalStateException("max page size < 1");
        }
        if (column_truncate_length.get() < 1) {
            throw new IllegalStateException("column truncate length < 1");
        }
        if (stats_truncate_length.get() < 1) {
            throw new IllegalStateException("stats truncate length < 1");
        }
    }
}

File: rust/sleeper_df/src/objects/ffi_sleeper_options.rs
//! Sleeper options configuration that clients may choose to alter. Defaults should be sensible for most clients.
/*
* Copyright 2022-2026 Crown Copyright
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

use std::ffi::{CStr, c_char};

pub const USE_READAHEAD_STORE: bool = false;
pub const READ_PAGE_INDEXES: bool = false;
pub const MAX_ROW_GROUP_SIZE: usize = 100_000;
pub const MAX_PAGE_SIZE: usize = 128 * 1024;
pub const COMPRESSION_CODEC: &CStr = c"zstd";
pub const WRITER_VERSION: &CStr = c"v2";
pub const COLUMN_TRUNCATE_LENGTH: usize = 128;
pub const STATS_TRUNCATE_LENGTH: usize = 2147483647;
pub const DICT_ENCODE_ROW_KEYS: bool = false;
pub const DICT_ENCODE_SORT_KEYS: bool = false;
pub const DICT_ENCODE_VALUES: bool = false;

/// Contains all the Sleeper options for DataFusion operation. These may not be needed for every DataFusion usage,
/// so come with reasonable defaults. See Java side documentation for explanation of fields.
///
/// *THIS IS A C COMPATIBLE FFI STRUCT!* If you updated this struct (field ordering, types, etc.),
/// you MUST update the corresponding Java definition in java/common/foreign-bridge/src/main/java/sleeper/foreign/datafusion/FFISleeperOptions.java.
/// The order and types of the fields must match exactly.
#[repr(C)]
pub struct FFISleeperOptions {
    pub use_readahead_store: bool,
    pub read_page_indexes: bool,
    pub max_row_group_size: usize,
    pub max_page_size: usize,
    pub compression: *const c_char,
    pub writer_version: *const c_char,
    pub column_truncate_length: usize,
    pub stats_truncate_length: usize,
    pub dict_enc_row_keys: bool,
    pub dict_enc_sort_keys: bool,
    pub dict_enc_values: bool,
}

impl Default for FFISleeperOptions {
    fn default() -> Self {
        Self {
            use_readahead_store: false,
            read_page_indexes: false,
            max_row_group_size: MAX_ROW_GROUP_SIZE,
            max_page_size: MAX_PAGE_SIZE,
            compression: COMPRESSION_CODEC.as_ptr(),
            writer_version: WRITER_VERSION.as_ptr(),
            column_truncate_length: COLUMN_TRUNCATE_LENGTH,
            stats_truncate_length: STATS_TRUNCATE_LENGTH,
            dict_enc_row_keys: DICT_ENCODE_ROW_KEYS,
            dict_enc_sort_keys: DICT_ENCODE_SORT_KEYS,
            dict_enc_values: DICT_ENCODE_VALUES,
        }
    }
}
